--- newt-0.52.14/form.c.strict_alias~	2012-09-22 15:02:33.944328303 +0200
+++ newt-0.52.14/form.c	2012-09-22 15:02:36.766362127 +0200
@@ -232,7 +232,7 @@ static int Gpm_Open(Gpm_Connect *conn, i
         goto err;
       } /*if*/
       strncpy (addr.sun_path, sock_name, sizeof (addr.sun_path));
-      if (bind (gpm_fd, (struct sockaddr*)&addr,
+      if (bind (gpm_fd, &addr,
                 sizeof (addr.sun_family) + strlen (addr.sun_path))==-1) {
         goto err;
       } /*if*/
@@ -242,7 +242,7 @@ static int Gpm_Open(Gpm_Connect *conn, i
       strcpy(addr.sun_path, GPM_NODE_CTL);
       i=sizeof(addr.sun_family)+strlen(GPM_NODE_CTL);
 
-      if ( connect(gpm_fd,(struct sockaddr *)(&addr),i)<0 )
+      if ( connect(gpm_fd,&addr,i)<0 )
         {
           struct stat stbuf;
 
